qcengine/programs/orca.py

```diff
@@ -0,0 +1,298 @@
+
+            input_file = "\n".join(input_file)
+        else:
+            # Some of the potential different template options
+            # (A) ordinary build_input (need to define a base template)
+            # (B) user wants to add stuff after normal template (A)
+            # (C) user knows their domain language (doesn't use any QCSchema quantities)
+
+            # # Build dictionary for substitute
+            # sub_dict = {
+            #     "method": input_model.model.method,
+            #     "basis": input_model.model.basis,
+            #     "charge": input_model.molecule.molecular_charge
+            # }
+
+            # Perform substitution to create input file
+            str_template = string.Template(template)
+            input_file = str_template.substitute()
+
+        return {
+            "commands": [which("orca"), "dispatch.mol"],
+            "infiles": {"dispatch.mol": input_file},
+            "scratch_directory": config.scratch_directory,
+            "input_result": input_model.copy(deep=True),
+        }
+
+    def parse_output(self, outfiles: Dict[str, str], input_model: "AtomicInput") -> "AtomicResult":
+
+        data = cclib.io.ccread(io.StringIO(outfiles["stdout"]))
+
+        properties = {}
+        extras = {}
+        extras["ev_to_hartrees"] = 0.0367493
+
+        # Process basis set data
+        properties["calcinfo_nbasis"] = data.nbasis
+        properties["calcinfo_nmo"] = data.nmo
+        properties["calcinfo_natom"] = data.natom
+        properties["scf_dipole_moment"] = data.moments[1].tolist()
+
+        # Grab the method from input
+        method = input_model.model.method.upper()
+
+        # Determining the final energy
+        # Throws an error if the energy isn't found for the method specified from the input_model.
+        try:
+            final_energy = data.scfenergies[-1] * extras["ev_to_hartrees"]
+        except:
+            raise KeyError(f"Could not find {method} total energy")
+
+        # Initialize output_data by copying over input_model.dict()
+        output_data = input_model.dict()
+
+        # Determining return_result
+        if input_model.driver == "energy":
+            output_data["return_result"] = final_energy
+            extras["CURRENT ENERGY"] = final_energy
+
+        elif input_model.driver == "gradient":
+            gradient = self.get_gradient(outfiles["outfiles"]["dispatch.engrad"])
+            output_data["return_result"] = gradient
+            extras["CURRENT ENERGY"] = final_energy
+            extras["CURRENT GRADIENT"] = gradient
+
+        # Final output_data assignments needed for the AtomicResult object
+
+        output_data["properties"] = properties
+        output_data["extras"].update(extras)
+        output_data["schema_name"] = "qcschema_output"
+        output_data["stdout"] = outfiles["stdout"]
+        output_data["success"] = True
+
+        return AtomicResult(**output_data)
+
+    def get_gradient(self, gradient_file):
```

Comment:
Can this be a private function with `_get_gradient`?

Reply:
Yes, no problem. Done.